src/components/editor/MonacoEditor.tsx

```diff
@@ -72,27 +78,48 @@ function MonacoEditor({
     //     // eslint-disable-next-line react-hooks/exhaustive-deps
     // }, [serverUpdate]);
 
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    const editorUpdateDebounce = useCallback(
-        debounce((val: string | undefined) => {
-            if (val && onChange) {
-                setIsChanging(true);
-                onChange(JSON.parse(val), path, currentCatalog.spec_type)
+    const updateValue = () => {
+        const currentValue = editorRef.current?.getValue();
+        if (onChange && currentValue) {
+            setStatus(EditorStatus.EDITING);
+
+            let parsedVal;
+            try {
+                parsedVal = JSON.parse(currentValue);
+            } catch {
+                setStatus(EditorStatus.INVALID);
+            }
+
+            if (parsedVal) {
+                setStatus(EditorStatus.SAVING);
+                onChange(
+                    parsedVal,
+                    currentCatalog.catalog_name,
+                    currentCatalog.spec_type
+                )
                     .then(() => {
-                        setIsChanging(false);
+                        setStatus(EditorStatus.SAVED);
                     })
                     .catch(() => {
-                        setIsChanging(false);
+                        setStatus(EditorStatus.SAVE_FAILED);
                     });
+            } else {
+                setStatus(EditorStatus.INVALID);
             }
-        }, 750),
-        []
-    );
+        }
+    };
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    const debouncedChange = useCallback(debounce(updateValue, 1000), [
+        currentCatalog,
+    ]);
 
     const handlers = {
-        change: (val: string | undefined) => {
-            setIsChanging(true);
-            editorUpdateDebounce(val);
+        change: () => {
+            if (status !== EditorStatus.EDITING) {
+                setStatus(EditorStatus.EDITING);
+            }
+            debouncedChange();
```

Comment:
If the status is already there don't want to hammer that set function because this handler is called on every letter.

This debounce change **does not work** for using actions/commands like undo and stuff. I'll start working on that soon.